### PR TITLE
feat: attempt DHCP discovery thrice

### LINKED ISSUE #1099: networkd: DHCP discovery should occur more than once
My (Intel, for what it's worth) network cards apparently report ready before they are _actually_ ready, which results in the `cli.Request()` to fail (no packet is actually emitted across the network).

In general, since a DHCP Request is a non-connection-oriented broadcast packet, it is not safe to assume that it was received (or sent, in this case), so it is a good idea to retransmit if there is no reply after the first one.

I would propose to retrying failed requests a limited number of times.  In a normal OS, this is typically forked into the background to retry a few times (or forever).

The problem I see presently is the DHCP fails after 10s, and the boot process moves on to download the userdata, which _does_ retry, but fails forever because DHCP is _not_ retried after the first failure.

### PR DESCRIPTION
**Pull request by Ulexus**
Fixes #1099

Signed-off-by: Seán C McCord <ulexus@gmail.com>

### CODE REVIEW

File: internal/app/networkd/pkg/address/dhcp.go [hunk 1]
```diff
@@ -158,12 +162,14 @@ func (d *DHCP) discover() (*dhcpv4.DHCPv4, error) {
...
-		// TODO: Make this a well defined error so we can make it not fatal
-		log.Println("failed dhcp request")
-		return nil, err
+	// TODO: this may be better handled by retrying indefinitely in the background
+	for i := 0; i < MaxRequests; i++ {
+		_, ack, err = cli.Request(context.Background(), mods...)
+		if err != nil {
+			// TODO: Make this a well defined error so we can make it not fatal
+			log.Println("failed dhcp request:", err)
+		}
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/address/dhcp.go]: Maybe we should do something like

`time.Sleep(i * time.Second)`

?

Comment (Ulexus) [reply internal/app/networkd/pkg/address/dhcp.go]: The dhcpv4 client internally waits a default of 10s for a timeout... but it's true that in the event of an immediate error, we shouldn't have a tight spin cycle.  I'll put an at-least timer in there.

Comment (andrewrynhard) [reply internal/app/networkd/pkg/address/dhcp.go]: Ahh wasn’t aware that the client already handled it to some degree. SGTM

Comment (Ulexus) [reply internal/app/networkd/pkg/address/dhcp.go]: Fixed

**Comment by andrewrynhard** [comment]
This working for you? We can merge if so.

**Comment by Ulexus** [comment]
Testing now

**Comment by Ulexus** [comment]
Yes, this appears to fix my observed problem.

**Comment by bradbeam** [comment]
Hold please. The library we use does three retries and is configurable with functional args.

**Comment by Ulexus** [comment]
Does three retries already, you are saying?

**Comment by Ulexus** [comment]
You appear to be correct:  the defaults are 5s timeout and 3 retries.  However, the process appears to timeout after on 10s, not 15, for me.  Perhaps it is triggering far earlier than expected.

**Comment by Ulexus** [comment]
In any case, this is the wrong solution, since the library includes the required tunables internally.

**Pull request closed without merging**